# Stage A Quiz Tag-Along code by Bamidele Tella

In [43]:
import pandas as pd
import numpy as np

In [44]:
df = pd.read_csv('https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv' , encoding = "latin-1")

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


In [46]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

In [47]:
df.corr()

<ipython-input-47-2f6f6606aa2c>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

In [48]:
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28



Given the following numpy array

array  = ([[94, 89, 63],

             [93, 92, 48],

             [92, 94, 56]])

How would you select  the elements in bold and italics from the array?

Options
array[ : 1, 1 : ]

array[ : 2, 0 : ]

array[ 1 : , : ]

array[ : 1, 1 :]

array[ : 2, 1 : ]


What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

Options
'l'

'e'

'i'

28


**'e'**

Which year had the least correlation with ‘Element Code’?

Options
2014

2015

2016

2017


In [41]:
corr = df.corr()
new_corr = corr['Element Code'].iloc[3:]
new_corr[new_corr==new_corr.min()]

<ipython-input-41-e02bd2400a92>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  corr = df.corr()


Y2016    0.023444
Name: Element Code, dtype: float64

Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

Options
Seychelles

Cabo Verde

Ghana

Guinea-Bissau

Comoros


In [56]:
selected_columns = df[['Area','Y2017']]

# Grouping by 'Area' and summing the 'Y2017' values
grouped = selected_columns.groupby('Area')['Y2017'].sum()
grouped[grouped==grouped.sort_values()[6]]

Area
Guinea-Bissau    19102.77
Name: Y2017, dtype: float64


How would you check for the number of rows and columns in a pandas DataFrame named df?

Options
df.shape( )

 len(df)

pd.length(df)

df.shape


In [ ]:
df.shape

(60943, 12)


Which of these python data structures is unorderly?

Options
Tuple

List

Dictionary

Set



**Set and Dictionary**

Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

Options
Kenya

South Africa

Egypt

Ethiopia

Nigeria


In [57]:
# Selecting columns 'Y2017' and 'Area'
selected_columns2 = df[['Y2017', 'Area']]

# Grouping by 'Area' and summing the 'Y2017' values
grouped2 = selected_columns2.groupby('Area')['Y2017'].sum()

# Finding the Area with the highest sum in 2017
area_with_highest_sum = grouped2.idxmax()

print("Area with the highest sum in 2017:", area_with_highest_sum)


Area with the highest sum in 2017: Nigeria



What is the total Protein supply quantity in Madagascar in 2015?

Options
174.91

357.12

173.05

180.29


In [ ]:
protein_madagascar_2015 = df[(df['Area'] == 'Madagascar') & (df['Element'] == 'Protein supply quantity (g/capita/day)')]

# Sum the 'Value' column to get the total Protein supply quantity
total_protein_supply_2015 = protein_madagascar_2015['Y2015'].sum()

print("Total Protein supply quantity in Madagascar in 2015:", total_protein_supply_2015)

Total Protein supply quantity in Madagascar in 2015: 173.04999999999998



What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

Options
10.60 and 17.01

8.54 and 9876.77

10.60 and 8.49

140.92 and 1671.86

135.23 and 1603.40


In [ ]:
print("Mean for the year 2017 is: ",df['Y2017'].mean())
print("Standard deviation for the year 2017 is: ",df['Y2017'].std())

Mean for the year 2017 is:  140.9177648602722
Standard deviation for the year 2017 is:  1671.8623590567995



What is the total number of unique countries in the dataset?

Options
45

39

50

49


In [ ]:
df['Area'].nunique()

49


If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = [‘Age’,’Nationality’,’Overall’]

How do you create a pandas DataFrame using this list, to look like the table below?

Options
pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])

pd.Dateframe(lst, column = [‘Age’,’Nationality’,’Overall’], index = [1,2,3])

pd.Dataframe(lst, columns = col, index = [1,2,3])

pd.datframe(lst, column = [‘Age’,’Nationality’,’Overall’],index = [1,2,3])

None of the above


In [ ]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ['Age','Nationality','Overall']

print(pd.DataFrame(lst, columns = col, index = [1,2,3]))
print(" ")
print(pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)]))

   Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92
 
   Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92



Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

Options
287134.67

292836.00

232118.43

3947019.23

3231287.75


In [ ]:
selected_columns = df[['Y2017', 'Element']]

# Grouping by 'Element' and summing the 'Y2017' values
grouped = selected_columns.groupby('Element')['Y2017'].sum()
print(grouped.Processing)

292836.0



Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

Options
Value Error

Type Error

Math Error

Standard Error


In [ ]:
my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

TypeError: ignored


Which of the following dataframe methods can be used to access elements across rows and columns?

Options
df.loc( : )

df.iloc[ : ]

df.loc( )

c and d

df.iloc( )


**B**


What is the total number and percentage of missing data in 2014 to 3 decimal places?

Options
1589 and 2.607%

1538 and 2.522%

1598 and 2.617%

1473 and 2.152 %

1506 and 2.477%


In [74]:
total = df['Y2014'].isnull().sum()
nrow = len(df['Y2014'])
percent = total / nrow * 100
print('Total Number of Missing Values: ', total)
print('Percentage of Missing Values: ', percent)

Total Number of Missing Values:  1589
Percentage of Missing Values:  2.6073544131401474



Which of the following is a python inbuilt module?

Options
Pandas

Seaborn

Math

Matplotlib



### Math

Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

How would you assign element 8 from the list to a variable x?

Options
x = y[1][-2]

x = y[1][-1]

x = y[1][1]

x = y[2][-1]


### x =y[1][1]


Answer the following questions based on the African food production dataset provided by the FAO website already provided

What is the total sum of Wine produced in 2015 and 2018 respectively?

Hint:

Perform a groupby sum aggregation on ‘Item’

Options
1584.08 and 1387.21

4251.81 and 4039.32

2134.34 and 3517.22

20675.72 and 26968.27


In [69]:
sc4 = df[['Item','Y2015','Y2018']]
grp4= sc4.groupby('Item').sum()
grp4.loc['Wine']

Y2015    4251.81
Y2018    4039.32
Name: Wine, dtype: float64


Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

Options
2015

2016

2014

2017

In [66]:
sc3 = df.groupby('Element').sum()
sc3.loc['Stock Variation'][3:].idxmax()

<ipython-input-66-2c0a6d6cf159>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  sc3 = df.groupby('Element').sum()


'Y2014'